src/project/git/StableDirectoryManager.ts

```diff
@@ -0,0 +1,135 @@
+import { registerShutdownHook } from "../../internal/util/shutdown";
+
+import { logger } from "../../internal/util/logger";
+
+import * as fs from "fs-extra";
+import * as os from "os";
+import { CloneDirectoryInfo, CloneOptions, DirectoryManager } from "./DirectoryManager";
+
+export interface StableDirectoryManagerOpts {
+
+    /**
+     * Maximum number of directories to use
+     * @type {number}
+     */
+    maxDirectories?: number;
+
+    /**
+     * Attempt to reuse directories?
+     */
+    reuseDirectories: boolean;
+
+    cleanOnExit?: boolean;
+
+    /**
+     * If provided, number of directories
+     */
+    baseDir: string;
+}
+
+const DefaultOpts = {
+
+    maxDirectories: 100,
+    cleanOnExit: true,
+};
+
+/**
+ * Build a stable working directory structure.
+ * Options determine whether it's cleared on exit.
+ */
+export class StableDirectoryManager implements DirectoryManager {
+
+    public opts: StableDirectoryManagerOpts;
+
+    private directoriesUsed = 0;
+
+    constructor(pOpts: StableDirectoryManagerOpts) {
+        this.opts = {
+            ...DefaultOpts,
+            ...pOpts,
+        };
+
+        logger.info("Setting up StableDirectoryManager with options %s", JSON.stringify(this.opts));
+
+        if (this.opts.cleanOnExit === true) {
+            logger.info("Registering shutdown hook to delete [%s]", this.opts.baseDir);
+            registerShutdownHook(() => {
+                logger.info("Cleaning up temporary directories under [%s]", this.opts.baseDir);
+                return fs.remove(this.opts.baseDir).then(() => 0);
+            });
+        }
+    }
+
+    public directoryFor(user: string, repo: string, branch: string, opts: CloneOptions): Promise<CloneDirectoryInfo> {
+        if (this.opts.reuseDirectories) {
+            // Attempt to reuse the directory
+            return this.existingDirectoryFor(user, repo, branch, opts)
+                .then(existing =>
+                    !!existing ? existing : this.freshDirectoryFor(user, repo, branch, opts));
+        } else {
+            return this.freshDirectoryFor(user, repo, branch, opts);
+        }
+    }
+
+    /**
+     * Return undefined if not found
+     */
+    private existingDirectoryFor(user: string, repo: string, branch: string,
+                                 opts: CloneOptions): Promise<CloneDirectoryInfo> {
+        const expectedPath = this.pathFor(user, repo);
+        return fs.pathExists(expectedPath)
+            .then(exists => {
+                if (exists) {
+                    logger.info("%s directories used of %s: Reusing path [%s]",
```

Comment:
wny would we care about that max on existing directories?

Reply:
It's just a useful piece of log info

Reply:
debug